File: lib/theProjector.js
// ----------

// THE PROJECTOR SCRAPER

// - parses data from https://theprojector.sg/films
// - extracts the following
    // event-list-item-module--cover-image--1H3q2 > picture
    // event-list-item-module--title--1r-44
    // rating-module--rating--3GlKm 
    // event-list-item-module--blurb--1Mh9e
    // event-list-item-module--theme--3QO97

// - dependancy installation
    // npm install playwright
    //  npx playwright install

// ----------

const { chromium } = require('playwright');

async function fetchAndParseHTML(url) {
    try {
        const browser = await chromium.launch();
        const page = await browser.newPage();
        await page.goto(url, { waitUntil: 'networkidle' });
        let previousHeight;
        let currentHeight = await page.evaluate(() => document.body.scrollHeight);
        while (true) {
            previousHeight = currentHeight;
            await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
            await page.waitForTimeout(3000); // wait for content to load
            currentHeight = await page.evaluate(() => document.body.scrollHeight);
            if (currentHeight === previousHeight) break;
        }
        const content = await page.content();
        await browser.close();
        return content;
    } catch (error) {
        console.error('Error fetching HTML:', error);
    }
}

async function extractFilms(url) {

    const htmlText = await fetchAndParseHTML(url);

    if (htmlText) {
        const { JSDOM } = require('jsdom');
        const dom = new JSDOM(htmlText);
        const doc = dom.window.document;

        function decodeHTMLEntity(encodedString) {
            const textarea = doc.createElement('textarea');
            textarea.innerHTML = encodedString;
            return textarea.value;
        }

        // -- debugging -- 
        // console.log(doc.title);
        // console.log(doc.body.innerHTML)
        // const mainBody = doc.querySelector('.event-list-module--events--3jC_9.event-list-module--grid--3HVrf');
        // console.log(mainBody.innerHTML);

        const projFilms = doc.querySelectorAll('.event-list-item-module--event-list-item--3aAfb.event-list-module--event-list-item--1YddO.undefined'); 
        console.log(`number of movies => ${projFilms.length}`);

        const projFilmsObject = [];

        if (projFilms.length) {
            projFilms.forEach(projFilm => {

                    // console.log(projFilm.innerHTML);

                    // event-list-item-module--cover-image--1H3q2 > picture
                    // event-list-item-module--categories--1-fI5
                    // event-list-item-module--title--1r-44
                    // rating-module--rating--3GlKm 
                    // event-list-item-module--blurb--1Mh9e
                    // event-list-item-module--theme--3QO97

                    const coverImageSrc = projFilm.querySelector('.event-list-item-module--cover-image--1H3q2 img')?.getAttribute('src') || 'N/A';
                    const title = projFilm.querySelector('.event-list-item-module--title--1r-44 a')?.innerHTML || 'N/A';
                    const rating = projFilm.querySelector('.rating-module--rating--3GlKm')?.innerHTML || 'N/A';
                    const rawBlurb = projFilm.querySelector('.event-list-item-module--blurb--1Mh9e')?.innerHTML || 'N/A';
                    const cleanBlurb = decodeHTMLEntity(rawBlurb)
                        ?.replace(/<\/?(b|em|strong|i|p)\b[^>]*>/gi, '') 
                        .replace(/<br\s*\/?>/gi, ' | ') 
                        .replace(/\s{2,}/g, ' ') 
                        .trim(); 
                    const theme = projFilm.querySelector('.event-list-item-module--theme--3QO97')?.innerHTML || 'N/A';

                    // console.log(`
                    //         -----
                    //         Title: ${title}
                    //         Description: ${cleanBlurb}
                    //         Rating: ${rating}
                    //         Theme: ${theme}
                    //         Poster src: ${coverImageSrc}
                    //         -----
                    //     `);

                    projFilmsObject.push({
                        title: title,
                        tags: theme,
                        description: cleanBlurb,
                        duration: null,
                        rating: rating,
                        bookTicketsUrl: "https://theprojector.sg/films",
                        posterSrc: coverImageSrc
                    });

            });
            return projFilmsObject;
        } else {
            console.log('Zero the projector films were found');
        }
    } else {
        console.log('HTML content not found');
    }
};

// ----- EXECUTION CODE -----
    // - for testing
    // - actual usage to be called from /src/main.js

// extractFilms("https://theprojector.sg/films").then((projFilmsObject) => {
//     if (projFilmsObject) {
//         console.log(JSON.stringify(projFilmsObject, null, 2));
//     } else {
//         console.log('no movies data were found');
//     }
// });

File: rub/theProjector.rb
# ----------

# THE PROJECTOR SCRAPER

# - parses data from https://theprojector.sg/films
# - extracts the following
    # event-list-item-module--cover-image--1H3q2 > picture
    # event-list-item-module--title--1r-44
    # rating-module--rating--3GlKm 
    # event-list-item-module--blurb--1Mh9e
    # event-list-item-module--theme--3QO97

# ----------

require 'selenium-webdriver'
require 'nokogiri'
require 'json'

def fetch_and_parse_html(url)
  driver = Selenium::WebDriver.for :chrome
  driver.navigate.to url

  previous_height = driver.execute_script('return document.body.scrollHeight')
  loop do
    driver.execute_script('window.scrollTo(0, document.body.scrollHeight)')
    sleep 3
    current_height = driver.execute_script('return document.body.scrollHeight')
    break if current_height == previous_height
    previous_height = current_height
  end

  content = driver.page_source
  driver.quit

  content
rescue => e
  puts "Error fetching HTML: #{e.message}"
  nil
end

def decode_html_entity(encoded_string)
  doc = Nokogiri::HTML::DocumentFragment.parse(encoded_string)
  doc.text
end

def extract_films(url)
  html_text = fetch_and_parse_html(url)
  return unless html_text

  doc = Nokogiri::HTML(html_text)

  proj_films = doc.css('.event-list-item-module--event-list-item--3aAfb.event-list-module--event-list-item--1YddO.undefined')

  proj_films_object = []

  if proj_films.any?
    proj_films.each do |proj_film|
      cover_image_src = proj_film.at_css('.event-list-item-module--cover-image--1H3q2 img')&.[]('src') || 'N/A'
      title = proj_film.at_css('.event-list-item-module--title--1r-44 a')&.text.strip || 'N/A'
      rating = proj_film.at_css('.rating-module--rating--3GlKm')&.text.strip || 'N/A'
      raw_blurb = proj_film.at_css('.event-list-item-module--blurb--1Mh9e')&.inner_html || 'N/A'
      clean_blurb = decode_html_entity(raw_blurb)
        .gsub(/<\/?(b|em|strong|i|p)\b[^>]*>/i, '')
        .gsub(/<br\s*\/?>/i, ' | ')
        .gsub(/\s{2,}/, ' ')
        .strip
      theme = proj_film.at_css('.event-list-item-module--theme--3QO97')&.text.strip || 'N/A'

      proj_films_object << {
        title: title,
        tags: theme,
        description: clean_blurb,
        duration: nil,
        rating: rating,
        book_tickets_url: "https://theprojector.sg/films",
        poster_src: cover_image_src
      }
    end

    puts "Number of TP movies => #{proj_films_object.length}"
    proj_films_object
  else
    puts 'Zero The Projector films were found'
    nil
  end
rescue => e
  puts "Error extracting films: #{e.message}"
  nil
end

# ----- EXECUTION CODE -----
    # For testing

films = extract_films("https://theprojector.sg/films")
if films
  puts JSON.pretty_generate(films)
else
  puts 'No movies data were found'
end